### ISSUE TITLE
ESLint enforce-module-boundaries rule fails with library split in secondary entry points and mixed loading scenarios

### ISSUE THREAD

**Issue by edbzn** [nrwl/nx#18552]
### Current Behavior

The ESLint rule enforce-module-boundaries enforce libs to be lazy loaded as far as a dynamic import is detected. This is great for most scenarios, but not working well when a library is split into multiple secondary entry points.

Consider the following scenario where **lib A** is split into 2 secondary entry points **core** and **feature**, and **lib B** needs to lazy load **feature** and to statically load **core**:

![image](https://github.com/nrwl/nx/assets/8522558/4f2ecbe0-f1ca-42e0-a0cb-102aca97aa86)

Then the enforce-module-boundaries rule is failing because lib A is marked as "lazy", and everything should be lazy loaded as far as a dynamic import is detected. 

```ts
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

/* Here is the line that throws "Static imports of lazy-loaded libraries are forbidden."
   but static import of the secondary entry point "core" should be allowed. */
import { CoreModule } from '@myorg/lib-a/core';

/* Only "feature" is lazy, not the whole library and all the secondary entry-points. */
async function getFeatureModule() {
  const  { FeatureModule } = await import('@myorg/lib-a/feature');
  return FeatureModule;
}

getFeatureModule();

@NgModule({
  imports: [CommonModule, CoreModule],
})
export class LibBModule {
}
```

### Expected Behavior

Statically importing **core** secondary entry points should not throw `static imports of lazy-loaded libraries are forbidden`.

Secondary entry points should be taken into account when checking for static imports when a library is considered lazy, allowing the rule to statically import secondary entry points that have not been imported dynamically.

### GitHub Repo

https://github.com/edbzn/nx-static-import-issue/tree/main

### Steps to Reproduce

1. Clone the [reproduction](https://github.com/edbzn/nx-static-import-issue)
2. Run `nx lint lib-b`
3. See the [line](https://github.com/edbzn/nx-static-import-issue/blob/e7c7dd1371a39f85edaf34143030d03c37fbb04a/libs/lib-b/src/lib/lib-b.module.ts#L6) that is failing

### Nx Report

```shell
nx 16.6.0
```

### Operating System

- [ ] macOS
- [x] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by meeroslav** [comment]
Thank you @edbzn for reporting this.

To be honest, this was on our list for a while. We will get to it as soon as possible.

**Comment by plutotv-aaa** [comment]
Thank you for working on this. Is there any update @meeroslav? Much appreciated.

**Comment by meeroslav** [comment]
Sadly not yet

**Comment by montella1507** [comment]
Hi i have come here to create issue when i saw i had been already created.

This is really common for shared libraries with secondary entry points as a code-splitting / tree-shaking implementation. 

We have several huge applications with libraries which are usually small and eagerly loaded (in HTML template of parent components, imported etc)...

however there are huge components too, which are usually loaded lazily (with custom dyanimc-component, dynamic import(), or... with the latest defer() angular feature). 

defer() is at the moment not possible to use as far as the component is usually loaded as secondary entry point...

@meeroslav  is it possible to escalate this issue, pllease?

**Comment by de-don** [comment]
+1

**Comment by montella1507** [comment]
Yeah, however this will lead to 40 exceptions in our case as far as we have /ui library with lot of components and secondary entry points as far as defer does not work effectively without that and we have lot of "3rd party"  huge libraries involved in them (spreadsheets etc.)

IT does not sound so "exceptional" then :D

**Comment by supernaut** [comment]
Not sure if this would cover it, but similarly I would love for the `@nx/enforce-module-boundaries` rule to be enforced on components in Next.js libraries with `'use client'` but not server components.

**Comment by wizardnet972** [comment]
Any updates?

**Comment by joarkm1** [comment]
I see that secondary entry points do not appear as separate nodes in nx graph. Maybe they should be?

**Comment by pvds** [comment]
> Thank you [@edbzn](https://github.com/edbzn) for reporting this.
> 
> To be honest, this was on our list for a while. We will get to it as soon as possible.

HI @meeroslav is this issue planned or on any roadmap?
Would be awesome if NX can also "understand" secondary entrypoints!

**Comment by montella1507** [comment]
@joarkm1 

> I see that secondary entry points do not appear as separate nodes in nx graph. Maybe they should be? 

Please no. Or opt-in. Every bigger monorepo has hundreds of SEPs...

**Comment by meeroslav** [comment]
Decided to fix everything in a single PR:
- Add support for package.json-based entry points
	- 	Include parsing of conditionals and nested exports
- Check if the static import belongs to the secondary entry point before raising the `Static imports of lazy-loaded libraries...` error

> Please note that right now, our graph does not track the exact import expression. We only store the project name and type of import. This limits us to covering all corner cases. For example, importing from the same secondary entry point both statically and dynamically would not be flagged as an error.
